pkg/prometheus/resource_selector.go

```diff
@@ -0,0 +1,577 @@
+// Copyright 2023 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/blang/semver/v4"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/assets"
+	"github.com/prometheus-operator/prometheus-operator/pkg/k8sutil"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+)
+
+type ResourceSelector struct {
+	l                    log.Logger
+	p                    monitoringv1.PrometheusInterface
+	store                *assets.Store
+	podMonitorLister     ListAllByNamespacer
+	serviceMonitorLister ListAllByNamespacer
+	probeLister          ListAllByNamespacer
+	namespaceInformers   cache.SharedIndexInformer
+	metrics              *operator.Metrics
+	accessor             *operator.Accessor
+}
+
+type ListAllByNamespacer interface {
+	ListAllByNamespace(namespace string, selector labels.Selector, appendFn cache.AppendFunc) error
+}
+
+func NewResourceSelector(l log.Logger, p monitoringv1.PrometheusInterface, store *assets.Store, podMonitorLister ListAllByNamespacer, serviceMonitorLister ListAllByNamespacer, probeLister ListAllByNamespacer, namespaceInformers cache.SharedIndexInformer, metrics *operator.Metrics) *ResourceSelector {
+	return &ResourceSelector{
+		l:                    l,
+		p:                    p,
+		store:                store,
+		podMonitorLister:     podMonitorLister,
+		serviceMonitorLister: serviceMonitorLister,
+		probeLister:          probeLister,
+		namespaceInformers:   namespaceInformers,
+		metrics:              metrics,
+		accessor:             operator.NewAccessor(l),
+	}
+}
+
+func (rs *ResourceSelector) SelectServiceMonitors(ctx context.Context) (map[string]*monitoringv1.ServiceMonitor, error) {
```

Comment:
nit: should we add a method comment outlining the general strategy for selecting the monitor? This method is so large that I feel it would help

Reply:
Could you elaborate on how this comment would look like?

Personally, I don't think a comment explaining a function step-by-step helps to make things simple. If the code changes, we often forget about the comment, which gets outdated.

Reply:
I was thinking of something to just allude to people to what is being done it doesn't need to be very descriptive. I'm not 100% sure about the last phrase.

```suggestion
// SelectServiceMonitors selects ServiceMonitors based on the selectors in the Prometheus CR and filters them
// returning only those with a valid configuration. This function also populates authentication stores.
func (rs *ResourceSelector) SelectServiceMonitors(ctx context.Context) (map[string]*monitoringv1.ServiceMonitor, error) {
```

Reply:
I feel like breaking this method into two separate ones would make more sense, making sure they do what they are named after 😅

With that said, I think this goes beyond what this PR is intended to change, so I'll just add the comment for now :)

Reply:
Yeah, I feel the same, but that's precisely why such a comment from my POV makes the difference here. Just by reading the name of the method, a developer will most probably think this method just selects ServiceMonitors when in reality it does at least 3 things. Select them, filter them and adds data to store.

I'm still not 100% sold that this should be separated, I would have to see how it turns out, but I agree that regardless this should be done in a follow-up